From 69d084784de196acec8ab703cd1b379af211d624 Mon Sep 17 00:00:00 2001
From: Artur Paszkiewicz <artur.paszkiewicz@intel.com>
Date: Fri, 22 Feb 2019 10:15:45 +0100
Subject: [PATCH] mdmon: don't attempt to manage new arrays when terminating
Git-commit: 69d084784de196acec8ab703cd1b379af211d624
Patch-mainline: mdadm-4.1-12
References: bsc#1127526

When mdmon gets a SIGTERM, it stops managing arrays that are clean. If
there is more that one array in the container and one of them is dirty
and the clean one is still present in mdstat, mdmon will treat it as a
new array and start managing it again. This leads to a cycle of
remove_old() / manage_new() calls for the clean array, until the other
one also becomes clean.

Prevent this by not calling manage_new() if sigterm is set. Also, remove
a check for sigterm in manage_new() because the condition will never be
true.

Signed-off-by: Artur Paszkiewicz <artur.paszkiewicz@intel.com>
Signed-off-by: Jes Sorensen <jsorensen@fb.com>
Signed-off-by: Coly Li <colyli@suse.de>
---
 managemon.c | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/managemon.c b/managemon.c
index 101231c..29b91ba 100644
--- a/managemon.c
+++ b/managemon.c
@@ -727,9 +727,7 @@ static void manage_new(struct mdstat_ent *mdstat,
 	dprintf("inst: %s action: %d state: %d\n", inst,
 		new->action_fd, new->info.state_fd);
 
-	if (sigterm)
-		new->info.safe_mode_delay = 1;
-	else if (mdi->safe_mode_delay >= 50)
+	if (mdi->safe_mode_delay >= 50)
 		/* Normal start, mdadm set this. */
 		new->info.safe_mode_delay = mdi->safe_mode_delay;
 	else
@@ -803,7 +801,7 @@ void manage(struct mdstat_ent *mdstat, struct supertype *container)
 				break;
 			}
 		}
-		if (a == NULL || !a->container)
+		if ((a == NULL || !a->container) && !sigterm)
 			manage_new(mdstat, container, a);
 	}
 }
-- 
2.16.4

